llvm/lib/Target/Mips/MipsAsmPrinter.cpp

```diff
@@ -329,6 +334,51 @@ static void emitDirectiveRelocJalr(const MachineInstr &MI,
   }
 }
 
+// NOTE: This is being used to control optimization of BALC in the linker.
+// TODO: Introduce a new relocation that will serve for this purpose only.
+static bool emitDirectiveRelocBalcStubs(const MachineInstr &MI,
+                                        MCContext &OutContext,
+                                        TargetMachine &TM,
+                                        MCStreamer &OutStreamer) {
+  assert(EmitNanoMipsBalcStubsReloc);
+  if (MI.getMF() && MI.getMF()->getFunction().hasOptSize()) {
```

Comment:
Note that we want to optimize calls under hasOptSize() and inhibit it for !hasOptSize()